daml-script/daml/Daml/Script.daml

```diff
@@ -0,0 +1,91 @@
+-- Copyright (c) 2019 The DAML Authors. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+daml 1.2
+module Daml.Script where
+
+import DA.Optional
+
+-- | A free monad
+data Free f a
+  = Pure a
+  | Free (f (Free f a))
+
+instance Functor f => Functor (Free f) where
+  fmap f (Pure a) = Pure (f a)
+  fmap f (Free x) = Free (fmap f <$> x)
+
+instance Functor f => Applicative (Free f) where
+  pure = Pure
+  Pure f <*> Pure a = Pure (f a)
+  Pure f <*> Free x = Free (fmap f <$> x)
+  Free x <*> my     = Free ((<*> my) <$> x)
+
+instance Functor f => Action (Free f) where
+  Pure a >>= f = f a
+  Free x >>= f = Free ((>>= f) <$> x)
+
+-- | A free applicative, since we don’t have existentials we have to use the weird RankNTypes encoding, this is isomorphic to
+-- forall b. Ap f b (Ap f (b -> a))
```

Comment:
```suggestion
-- forall b. Ap (f b) (Ap f (b -> a))
```

Reply:
Good catch, thanks! fixed in https://github.com/digital-asset/daml/pull/3438